CalibPPS/AlignmentRelative/src/LocalTrackFitter.cc

```diff
@@ -0,0 +1,223 @@
+//----------------------------------------------------------------------------------------------------
+
+LocalTrackFitter::LocalTrackFitter(const edm::ParameterSet &ps)
+    : verbosity(ps.getUntrackedParameter<unsigned int>("verbosity", 0)),
+      minimumHitsPerProjectionPerRP(ps.getParameter<unsigned int>("minimumHitsPerProjectionPerRP")),
+      maxResidualToSigma(ps.getParameter<double>("maxResidualToSigma")) {}
+
+//----------------------------------------------------------------------------------------------------
+
+bool LocalTrackFitter::fit(HitCollection &selection, const AlignmentGeometry &geometry, LocalTrackFit &trackFit) const {
+  if (verbosity > 5)
+    printf(">> LocalTrackFitter::Fit\n");
+
+  bool selectionChanged = true;
+  unsigned int loopCounter = 0;
+  while (selectionChanged) {
+    // fit/outlier-removal loop
+    while (selectionChanged) {
+      if (verbosity > 5)
+        printf("* fit loop %u\n", loopCounter++);
+
+      bool fitFailed = false;
+      fitAndRemoveOutliers(selection, geometry, trackFit, fitFailed, selectionChanged);
+
+      if (fitFailed) {
+        if (verbosity > 5)
+          printf("\tFIT FAILED\n");
+        return false;
+      }
+    }
+
+    // remove pots with too few active planes
+    removeInsufficientPots(selection, selectionChanged);
+  }
+
+  return true;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void LocalTrackFitter::fitAndRemoveOutliers(HitCollection &selection,
+                                            const AlignmentGeometry &geometry,
+                                            LocalTrackFit &trackFit,
+                                            bool &failed,
+                                            bool &selectionChanged) const {
+  if (verbosity > 5)
+    printf(" - LocalTrackFitter::FitAndRemoveOutliers\n");
+
+  if (selection.empty()) {
+    failed = true;
+    return;
+  }
+
+  // check if input size is sufficient
+  if (selection.size() < 4) {
+    failed = true;
+    return;
+  }
+
+  // build matrices and vectors
+  TMatrixD A(selection.size(), 4);
+  TMatrixD Vi(selection.size(), selection.size());
+  TVectorD measVec(selection.size());
+  unsigned int j = 0;
+  for (auto it = selection.begin(); it != selection.end(); ++it, ++j) {
+    const unsigned int &detId = it->id;
+
+    const DetGeometry &d = geometry.get(detId);
+    const auto &dirData = d.getDirectionData(it->dirIdx);
+
+    A(j, 0) = it->z * dirData.dx;
+    A(j, 1) = dirData.dx;
+    A(j, 2) = it->z * dirData.dy;
+    A(j, 3) = dirData.dy;
+
+    measVec(j) = it->position + dirData.s - (it->z - d.z) * dirData.dz;  // in mm
+
+    Vi(j, j) = 1. / it->sigma / it->sigma;
+  }
+
+  // evaluate local track parameter estimates (h stands for hat)
+  TMatrixD AT(4, selection.size());
+  AT.Transpose(A);
+  TMatrixD ATViA(4, 4);
+  ATViA = AT * Vi * A;
+  TMatrixD ATViAI(ATViA);
+
+  try {
+    ATViAI = ATViA.Invert();
+  } catch (...) {
```

Comment:
Why have you not used the following?
```suggestion
  } catch (cms::Exception& e) {
```
I think that what happens is that the CMSSW framework handles the ROOT errors and throws them as `cms::Exception`. ROOT itself doesn't throw exceptions in these cases. If it is rare to have non-invertible matrices here, I think it's fine to catch the exceptions. But if non-invertible matrices occur regularly, we may not want to have the overhead of throwing an exception.

Reply:
Thanks @christopheralanwest , I've applied your suggestion! Furthermore, I checked with the real-life example in `CalibPPS/AlignmentRelative/test/test_with_data` that the catch block is invoked in less than ~1% calls to `LocalTrackFitter::fitAndRemoveOutliers`. In addition note that this code represents a special calibration that would be run "manually" (i.e. NOT in any central WF) only a couple of times per year.